### ISSUE TITLE
VMs broken on boot (ConnectionRefused)

### ISSUE THREAD

**Issue by pwaller** [astro/microvm.nix#312]
[note, this issue possibly describes two problems; I consider the one on nixpkgs@master + astro/microvm@master to be the one of greater relevance]

Probably ok nixpkgs rev: (2024-12-08) nixos/nixpkgs@3d58debbb538678cf297954e520d4a4ea3ff2e15
Probably bad nixpkgs rev: (2024-12-09) nixos/nixpkgs@b356fafe02dd1d01d58153ae02326268ed584a51
microvm is held at (2024-09-06) caac7808d1e31f8a0fa408338cd3736947cb226d

Platform is x86_64-linux; both host and VMs are running the same input nixpkgs.

After the move from OK to BAD nixpkgs revision above, microvms fail on boot with a message like the following:

```
cloud-hypervisor: 61.191516s: <vmm> ERROR:virtio-devices/src/vhost_user/vu_common_ctrl.rs:412 -- \
Failed connecting the backend after trying for 1 minute: \
VhostUserProtocol(SocketConnect(Os { code: 111, kind: ConnectionRefused, message: "Connection refused" }))
```

I'm using declarative VMs with cloud-hypervisor.

My method of testing for functioning/non-function is to do `nixos-rebuild boot` followed by `microvm -u`, followed by a reboot.

The observed behaviour is that everything restarts repeatedly during boot, and the VMs do not come up.

I have subsequently tried to update everything including astro/microvm to a commit from today, but that seems broken too:

```
nix flake update \
  --override-input microvm github:astro/microvm.nix/3910e65c3d92c82ea41ab295c66df4c0b4f9e7b3 \
  --override-input nixpkgs github:NixOS/nixpkgs/0b04b1f3548f38d045f46821d8f41cab5680fed7 \
  nixpkgs microvm
```

In this new configuration, cloud hypervisor log looks like the below, with no indication of the machine starting to boot (via console messages).

```
Dec 13 12:57:56 foo microvm@foobar[46704]: cloud-hypervisor: 19.556438ms: <vmm> WARN:net_util/src/open_tap.rs:105 -- Tap vm-foobar already exists. IP configuration will not be overwritten.
Dec 13 12:57:58 foo microvm@foobar[46704]: cloud-hypervisor: 1.800710s: <vmm> WARN:net_util/src/open_tap.rs:105 -- Tap vm-foobar already exists. IP configuration will not be overwritten.
Dec 13 12:57:59 foo microvm@foobar[46704]: cloud-hypervisor: 3.132105s: <vmm> WARN:net_util/src/open_tap.rs:105 -- Tap vm-foobar already exists. IP configuration will not be overwritten.
Dec 13 12:58:01 foo microvm@foobar[46704]: cloud-hypervisor: 4.897196s: <vmm> WARN:net_util/src/open_tap.rs:105 -- Tap vm-foobar already exists. IP configuration will not be overwritten.
Dec 13 12:58:03 foo microvm@foobar[46704]: cloud-hypervisor: 6.567278s: <vmm> WARN:net_util/src/open_tap.rs:105 -- Tap vm-foobar already exists. IP configuration will not be overwritten.
Dec 13 12:58:05 foo microvm@foobar[46704]: cloud-hypervisor: 8.372956s: <vmm> WARN:net_util/src/open_tap.rs:105 -- Tap vm-foobar already exists. IP configuration will not be overwritten.
Dec 13 12:58:06 foo microvm@foobar[46704]: cloud-hypervisor: 9.705110s: <vmm> WARN:net_util/src/open_tap.rs:105 -- Tap vm-foobar already exists. IP configuration will not be overwritten.
```

Diffing the closures between the original functioning system and non-functioning system shows very little of interest. I see: `linux: 6.6.63, 6.6.63-modules → 6.6.64, 6.6.64-modules, +55.6 KiB`.

Both are using the same cloud-hypervisor (`/nix/store/9s494x3d8drhwl3vs1abnbhx8djmly7h-cloud-hypervisor-42.0`).

**Comment by pwaller** [comment]
The following flake describing a host + guest is enough to reproduce.

Fails:
```
nix flake update --override-input microvm github:astro/microvm.nix/3910e65c3d92c82ea41ab295c66df4c0b4f9e7b3   --override-input nixpkgs github:NixOS/nixpkgs/0b04b1f3548f38d045f46821d8f41cab5680fed7
```

Succeeds:
```
nix flake update --override-input microvm github:astro/microvm.nix/caac7808d1e31f8a0fa408338cd3736947cb226d   --override-input nixpkgs github:NixOS/nixpkgs/b1dd465e8139748a8e26037fdd4c5ffe79457cbd
```

----

# Reproducer

```
#  Run a flake update command from above to reproduce good or bad configuration)
rm -f simple-microvm-host.qcow2
time nix build .#nixosConfigurations.simple-microvm-host.config.system.build.vm -L
QEMU_KERNEL_PARAMS="console=ttyS0 systemd.journald.forward_to_console=1" \
    ./result/bin/run-simple-microvm-host-vm -nographic
```

This is enough to show the problem on the console. In the working case, you can see the guest boots, in the failing case it does not. Deleting `microvm.interfaces` gets rid of the spammy `Tap vm-foobar already exists. IP configuration will not be overwritten` warnings, but instead leaves it spinning in the 'start microvm' state with no following output at all.

# `flake.nix`

```nix
{
  inputs = {
    microvm.url = "github:astro/microvm.nix";
    microvm.inputs.nixpkgs.follows = "nixpkgs";
    nixpkgs.url = "nixpkgs/nixos-unstable-small";
  };

  outputs = inputs @ {
    microvm,
    nixpkgs,
    self,
    ...
  }: {
    nixosConfigurations.simple-microvm-host = nixpkgs.lib.nixosSystem {
      system = "x86_64-linux";
      modules = [
        microvm.nixosModules.host
        {
          networking.hostName = "simple-microvm-host";
          system.stateVersion = "24.11";
          services.getty.autologinUser = nixpkgs.lib.mkForce "root";
          microvm.vms.simple-guest.flake = self;
        }
      ];
    };
    nixosConfigurations.simple-guest = nixpkgs.lib.nixosSystem {
      system = "x86_64-linux";
      modules = [
        microvm.nixosModules.microvm
        {
          microvm.hypervisor = "cloud-hypervisor";
          system.stateVersion = "24.11";

          networking.hostName = "simple-guest";
          microvm.interfaces = [
            {
              type = "tap";
              id = "vm-simple-guest";
              mac = "02:00:00:00:00:02";
            }
          ];
        }
      ];
    };
  };
}
```

**Comment by pwaller** [comment]
With astropy @ 3910e65c3d92c82ea41ab295c66df4c0b4f9e7b3
This bisects on nixpkgs to nixos/nixpkgs@d76dafcd29ff607b7963fa76a68b92b7cf94de21, which is the linux bump: `linux: 6.6.63 → 6.6.64`.

**Comment by JohannesEbke** [comment]
Reporting for @pwaller,  we've now bisected this into the kernel update 6.6.63 to 6.6.64 to commit [83d123e27623713dd69eed2569eacf5f1b3c9033](https://github.com/gregkh/linux/commit/83d123e27623713dd69eed2569eacf5f1b3c9033) of the linux kernel.

**Comment by pwaller** [comment]
The patches [were reverted in 6.6.65](https://cdn.kernel.org/pub/linux/kernel/v6.x/ChangeLog-6.6.65), and nixpkgs master is now at 6.6.66 https://github.com/NixOS/nixpkgs/pull/365310.


I've confirmed the reproducer is functioning correctly at this nixpkgs commit.

**Comment by SuperSandro2000** [comment]
Should we start to maintain a list of bad kernel versions? Iirc 6.11 also recently had a regression...

**Comment by pwaller** [comment]
Kernel fix has now landed on nixos-unstable-small and tested in the environment the problem was originally observed on, so I think this issue can be closed?

**Comment by RooSoft** [comment]
> Dec 13 12:58:06 foo microvm@foobar[46704]: cloud-hypervisor: 9.705110s: <vmm> WARN:net_util/src/open_tap.rs:105 -- Tap vm-foobar already exists. IP configuration will not be overwritten.

Experiencing some similar problems as we speak with cloud hypervisor and I have no idea how to fix it. Currently running the latest version of microvm.nix...

**Comment by pwaller** [comment]
@RooSoft 

For the problem described in this issue:

* For release-24.11 you need a commit before (https://github.com/NixOS/nixpkgs/pull/363872/commits/c5156287d704888daac8d6f46a1e26086b2a36e7) or after (https://github.com/NixOS/nixpkgs/commit/c2188abafab17076064b63a4f783eeb02485f8c1)

* For nixos-unstable you need a commit before (https://github.com/NixOS/nixpkgs/commit/d10eccec9d957b19cb08bee257808c9f231080ae) or after (https://github.com/NixOS/nixpkgs/commit/8e293569c3e552c996eb9ea3a19b1229cc684da6).

In each case you can check with a local checkout of nixpkgs if your chosen commit is an ancestor or not:

* `git merge-base --is-ancestor <before commit from above> HEAD && echo has problem`  
* `git merge-base --is-ancestor <after commit from above> HEAD && echo has fix`  

Currently, both nixos-unstable and nixos-24.11 have the problem in it but not the fix. So you'll need to revert back to the parent of the commit which introduced the problem or sooner.

**Comment by RooSoft** [comment]
thanks @pwaller... so, if I understand well, it's not gonna work with 24.11 anytime soon, right?

**Comment by pwaller** [comment]
The fix has been backported onto 24.11, per the above comment (you need a commit after https://github.com/NixOS/nixpkgs/commit/c2188abafab17076064b63a4f783eeb02485f8c1). So just as the brokenness came to your branch (it was backported there), so will the fix.

release-24.11 is built regularly here: https://hydra.nixos.org/jobset/nixos/release-24.11

The most recently built commit looks young enough it should work. Just make sure you're using the release-24.11 branch and not the nixos-24.11 branch.

For reasons I don't fully understand, the most recent commit listed on hydra for the release branch is actually behind the current release-24.11 branch (and the nixos-24.11 branch is behind the most recent hydra commit).

**Comment by pwaller** [comment]
https://nixos.wiki/wiki/FAQ#Nixpkgs_branches

Ah, you want the nixos-24.11 branch; which reflects what hydra has most recently built. So you could take the release-24.11 branch but it will involve building more for yourself.
